embassy-stm32/src/uid.rs

```diff
@@ -0,0 +1,33 @@
+use core::sync::atomic::{AtomicBool, Ordering};
+
+/// Get this device's unique 96-bit ID.
+pub fn uid() -> &'static [u8; 12] {
+    unsafe { &*crate::pac::UID.uid(0).as_ptr().cast::<[u8; 12]>() }
+}
+
+/// Get this device's unique 96-bit ID, encoded into a string of 24 hexadecimal ASCII digits.
+pub fn uid_hex() -> &'static str {
+    unsafe { core::str::from_utf8_unchecked(uid_hex_bytes()) }
+}
+
+/// Get this device's unique 96-bit ID, encoded into 24 hexadecimal ASCII bytes.
+pub fn uid_hex_bytes() -> &'static [u8; 24] {
+    const HEX: &[u8; 16] = b"0123456789ABCDEF";
+    static mut UID_HEX: [u8; 24] = [0; 24];
+    static LOADED: AtomicBool = AtomicBool::new(false);
+    if !LOADED.load(Ordering::Acquire) {
```

Comment:
You need atomic CAS for this to be sound, otherwise you can have htis:

- thread A: sees LOADED=false
- thread B: sees LOADED=false
- thread B: writes to UID_HEX
- thread B: returns
- thread B: user code starts to read from UID_HEX
- thread A: writes to UID_HEX
- One thread writing and another reading without synchronization -> UB

the CS doesn't save you there, the semantics for `critical_section` are "no other CS can run concurrently", not "no other code can run concurrently". (the cortex-m single-core impl does guarantee the latter, but not necessarily all do)

Since CAS is annoying on M0's I'd just make LOADED a `static mut` too and access it inside the CS. A bit slower because you now have to take the CS if already loaded, but i'd say it's fine because this operation should be rare.

Reply:
Good point, I've swapped as you suggested. Do you still prefer this over just computing it and returning an array directly?

Reply:
it's definitely a bit odd, but it's convenient to be able to get a `&str` and you only get the memory usage if you actually use it, so there's little downside I guess?